spring-cloud-aws-sns/src/test/java/io/awspring/cloud/sns/integration/SnsTemplateIntegrationTest.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2013-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.awspring.cloud.sns.integration;
+
+import static org.testcontainers.containers.localstack.LocalStackContainer.Service.DYNAMODB;
+import static org.testcontainers.containers.localstack.LocalStackContainer.Service.SNS;
+
+import io.awspring.cloud.sns.Person;
+import io.awspring.cloud.sns.core.SnsNotification;
+import io.awspring.cloud.sns.core.SnsTemplate;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.testcontainers.containers.localstack.LocalStackContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.sns.SnsClient;
+import software.amazon.awssdk.services.sns.model.CreateTopicRequest;
+
+/**
+ * Integration tests for {@link SnsTemplate}.
+ *
+ * @author Matej Nedic
+ * @since 3.0
+ */
+@Testcontainers
+class SnsTemplateIntegrationTest {
+	private static final String TOPIC_NAME = "my_topic_name";
+	private static SnsTemplate snsTemplate;
+	private static SnsClient snsClient;
+
+	@Container
+	static LocalStackContainer localstack = new LocalStackContainer(
+			DockerImageName.parse("localstack/localstack:0.14.0")).withServices(SNS).withReuse(true);
+
+	@BeforeAll
+	public static void createTable() {
+		snsClient = SnsClient.builder().endpointOverride(localstack.getEndpointOverride(DYNAMODB))
+				.region(Region.of(localstack.getRegion()))
+				.credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create("noop", "noop")))
+				.build();
+		snsTemplate = new SnsTemplate(snsClient, null);
+	}
+
+	@Test
+	void send_validTextMessage_usesTopicChannel_auto_create() {
+		snsTemplate.convertAndSend(TOPIC_NAME, "message");
+	}
+
+	@Test
+	void send_validTextMessage_usesTopicChannel_send_arn() {
+		String topic_arn = snsClient.createTopic(CreateTopicRequest.builder().name(TOPIC_NAME).build()).topicArn();
+		snsTemplate.convertAndSend(topic_arn, "message");
```

Comment:
We should verify that the notification we sent is correct. The same for other test.

Reply:
When you say correct you mean that Request is valid? Not sure I follow you on this. Since making a subscription and listening to it will require us to autoconfigure endpoint and assert when message comes. If I understood this is what you ask?

Reply:
Yes pretty much. So an example:

```java
@Test
void send_validTextMessage_usesTopicChannel_send_arn() {
	String queueUrl = sqsClient.createQueue(r -> r.queueName("my-queue")).queueUrl();
	String topicArn = snsClient.createTopic(CreateTopicRequest.builder().name(TOPIC_NAME).build()).topicArn();
	snsClient.subscribe(r -> r.topicArn(topicArn).protocol("sqs").endpoint(queueUrl));

	snsTemplate.convertAndSend(topicArn, "message");

	ObjectMapper objectMapper = new ObjectMapper();

	await().untilAsserted(() -> {
		ReceiveMessageResponse response = sqsClient.receiveMessage(r -> r.queueUrl(queueUrl));
		assertThat(response.hasMessages()).isTrue();
		JsonNode body = objectMapper.readTree(response.messages().get(0).body());
		assertThat(body.get("Message").asText()).isEqualTo("message");
	});
}
```
We don't need to verify all the properties of the retrieved message as we test it anyway in unit test, but I believe we should verify that the full cycle works. If we don't, and leave it as it is I think we could also just drop this integration test completely.